---
title: "<Protocol name> ICS Processing"
shorttitle: "<Protocol name> ICS"
from:
 - name: Your Name
   email: <your_id>@fredhutch.org
includesummary: yes
summary:
  ICS processing for the <Protocol name> study
---

<!--
Program Name: preprocess_flow.Rmd
Creation Date: 1/1/2010
Full name of author: Your Name
Project or Protocol: <Protocol name>
Purpose or description of program: ICS data processing
Location of program: git@github.com/FredHutch/<protocol name>.git
Location of input data: git@github.com/FredHutch/, raw data at `/trials/cavd/obj1/cvdNNN/`
-->

This report documents the <Protocol name> ICS data processing. We first load FlowWorkspace and its dependencies, and we define input and scratch locations. The raw data is very large (close to 30 GB), and parsing it is time-consuming. So we parse it once and save the parsed data to the local machine in a scatch directory. All subsequent processing can be performed locally on the parsed data, and reprocessing is substantially faster if that parsed data is retained until the package is finalized.

```{r prelim}
# Run-specific parameters
parseXML <- TRUE # set to TRUE in order to parse XML from raw files, FALSE otherwise
runMIMOSA <- TRUE # Run MIMOSA and make response call?
runCOMPASS <- TRUE # Run COMPASS to create COMPASS object and compute polyfunctionality scores?

PATH <- "/Volumes/trials/cavd/obj1/cvdNNN" # Location of fcs files on network drive
SCRATCH <- file.path("~", "scratch", pkgName) # scratch path
kw <- c("Animal ID", "Group", "Stim", "Time Point", "CYTNUM", "User", "Run", "WELL ID", "SETTINGS") # relevant keywords found in XML files
Stims <- c("FIV-GAG", "SIV-GAG", "HIV-ENV", "RSV-F")

suppressWarnings(
  suppressPackageStartupMessages({
    library(knitr)
    library(rmarkdown)
    library(ggplot2)
    library(devtools)
    library(xtable)
    library(flowWorkspace)
    library(openCyto)
    library(stringr)
    library(here)
    library(tidyverse)
    library(ggcyto)
    library(flowIncubator)
    library(data.table)
  }))

# Functions
sys.source("functions.R", envir=topenv())
sys.source("qc_functions.R", envir=topenv())
sys.source('protocol_specific.R', envir=topenv())

# create plots directory under vignettes/
plotsdir <- file.path("..", "vignettes", "plots")
if(!dir.exists(plotsdir)){
  dir.create(plotsdir, recursive = TRUE, mode = "0755")
}

options(xtable.comment = FALSE, datatable.verbose = FALSE, scipen = 10)
opts_chunk$set(results = "asis", cache = FALSE, echo = TRUE, tidy = TRUE, messages = TRUE, warning = TRUE,
               dev = c("pdf", "png"), dpi = 300, fig.pos = 'htbp', fig.height = 9, fig.width = 7.5, scipen=0, digits=4)
```

We now parse the XML files to identify the workspaces, then we select which groups we want to parse from the workspaces. Despite its name, we do not want the group called 'All Samples'; instead, we want the group called 'samples'. However, some workspaces do not have such a group. In that case, we use the group(s) that refer to bottom plates. There may be more than one, in which case we need to process each workspace once for each group we want to extract.

The keywords argument to `parseWorkspace()` requires examining one or more XML files, usually around line 250. Keywords in lower case tend to be the ones added by the experimenter(s).

```{r processXML, eval=parseXML}

# delete temp files if they exist from a previous run
if(dir.exists(file.path(SCRATCH, "gs"))){
  unlink(file.path(SCRATCH, "gs"), recursive = TRUE, force = TRUE)
}

workspaces <- list.files(path = PATH,
                         pattern = "xml",
                         full.names = TRUE)

ws <- list()
for (i in workspaces) {
  ws[[i]] = openWorkspace(i)
}

# Find all "Samples" groups
# this usually works but may not always work
# You want the function to identify the sample group called "Samples" or "samples" in each workspace, but sometimes there isn't one and
# sometimes there is more than one with a name similar to "samples." Note that you rarely if ever want the "ALL SAMPLES" group.
# If it's unclear, you may need to examine each one, check with Greg, or check with the lab to figure out which one(s) you need to parse.
l <- lapply(ws, function(x) levels(getSampleGroups(x)$groupName))
findfun <- function(x){
  ifelse(length(which(toupper(x)=="SAMPLES")) > 0, which(toupper(x)=="SAMPLES"), which(toupper(x) != "ALL SAMPLES" & grep("BOTTOM", toupper(x))))
}
groupnames <- do.call(rbind, lapply(l, findfun))

# Parse all workspaces into a list
gs = list()
for (i in seq_along(ws)) {
  gs[[i]] = parseWorkspace(ws[[i]],
                          name = groupnames[i, ],
                          keywords = kw,
                          path = PATH)
}

# Reconcile different gating structures & remove redundant nodes
gs_groups <- groupByTree(gs)
checkRedundantNodes(gs_groups, silent=TRUE)

toRm <- setdiff(getNodes(gs_groups[[1]][[1]]), getNodes(gs_groups[[2]][[1]]))
for(node in toRm)
  for(i in 1:length(gs_groups[[1]]))
    Rm(node, gs_groups[[1]][[i]])

# Did it work?
checkRedundantNodes(gs_groups, silent=TRUE)
gs_groups <- groupByTree(gs)
length(gs_groups) # 1
```

Now we check that the marker names are all consistent. These will need to be updated if not consistent. Otherwise `getSingleCellExpression()` will fail.

```{r standardize_and_save, eval = parseXML}
all_sample_markers=list()
for(i in sampleNames(gs)){
  all_sample_markers[[i]] = markernames(gs@data[i])
}

new_markernames <- parameters(gs@data[[1]])@data$desc
channelnames <- parameters(gs@data[[1]])@data$name
names(new_markernames) <- channelnames
new_markernames <- na.omit(new_markernames)
attr(new_markernames, "na.action") = NULL
markernames(gs) <- new_markernames

gsl <- GatingSetList(gs) # convert list of GatingSets into a GatingSetList.
gs <- rbind2(gsl) # convert GatingSetList into a GatingSet.

save_gs(gs, path <- file.path(SCRATCH, "gs"), cdf = "link")
```


```{r 'gating_hierarchy', eval = parseXML}
nodes <- getNodes(gs)
print(nodes)
png(filename = file.path("..", "vignettes", "plots", "gating_hierarchy.png")
plot(gs)
dev.off()
```

Next we add the boolean gates. The updated GatingSet containing booleans is also saved to the scratch directory.

```{r BooleanGating, eval=parseXML}
gs <- load_gs(path=file.path(SCRATCH, "gs"))

# delete boolean temp files if they exist from a previous run
if(dir.exists(file.path(SCRATCH, "gs_booleans"))){
  unlink(file.path(SCRATCH, "gs_booleans"), recursive = TRUE, force = TRUE)
}

# Boolean gating
# Next we need to add the boolean gates beneath the total memory populations.
# this function is customized to the standard gating we've been getting from the Foulds lab. It will need to be updated if the gating differs.
gatingargs <- function(gatingset, tsub){
  nodes <- getNodes(gatingset)
  x <- gsub("/singlets/live/cd3/", "", nodes[grepl("totm/", nodes, fixed=TRUE)])
   x1 <- x[grepl(tsub, x)]
   x1 <- paste(x1, collapse=":")
   print(x1)
   return(x1)
}

cd4gatingargs <- gatingargs(gs, "cd4")
cd8gatingargs <- gatingargs(gs, "cd8")
openCyto:::add_pop(gs, gating_method = "polyFunctions", parent = "cd4/totm", gating_args = cd4gatingargs)
openCyto:::add_pop(gs, gating_method = "polyFunctions", parent = "cd8/totm", gating_args = cd8gatingargs)

# Save
save_gs(gs, path=file.path(SCRATCH, "gs_booleans"), cdf = "link")
```

We now extract the ICS data and computed derived variables.

```{r extractICS}
# Population statistics
gs <- load_gs(path=file.path(SCRATCH, "gs_booleans"))

# We extract the population statistics
ICS_data <- getPopStats(gs, path="full", bool=TRUE)
ICS_data[, HierarchyLevel := str_count(Population, "/")]

# merge with pdata
# all the gsubs below assume standard Foulds gating and will need to updated if gating differs
pd <- pData(gs)
pd$name <- rownames(pd)
ICS_data <- merge(ICS_data, pd, by="name")
ICS_data[, Population := gsub("/singlets/live/cd3/", "", Population)]
ICS_data[, Population := gsub("/singlets/live/", "", Population)]
ICS_data[, Population := gsub("/singlets/", "", Population)]
ICS_data[, Population := gsub("/singlets", "singlets", Population)]
ICS_data[, Parent := gsub("/singlets/live/cd3/", "", Parent)]
ICS_data[, Parent := gsub("/singlets/live/", "", Parent)]
ICS_data[, Parent := gsub("/singlets/", "", Parent)]
ICS_data[, Parent := gsub("/singlets", "singlets", Parent)]
ICS_data[, Population := gsub(paste0(Parent, "/"), "", Population), .(name, Parent)]
ICS_data[, Population := gsub("IL-", "IL", gsub("IFN-G", "IFNg", toupper(Population)))]
ICS_data[, Population := gsub("IL17", "IL17A", Population, fixed=TRUE)]
ICS_data[, Parent := gsub("IL-", "IL", gsub("IFN-G", "IFNg", toupper(Parent)))]
ICS_data[, Parent := toupper(Parent)]

# Standardize data values
ICS_data[, `:=`(Stim = ifelse(Stim==ICS_control, Stim, toupper(Stim)),
                obstype = ifelse(grepl(":", Population, fixed=TRUE), "Boolean",
                                 ifelse(grepl("TOTM", Parent) | Parent == "CD4/CM/CXCR5", "Provided Marginal", "Hierarchical Subset")))]

# Parse Boolean nomenclature in raw data and make it more human readable
ICS_data[, tmp := gsub(":", "", gsub(gsub("/", ":", Parent, fixed=TRUE), "", Population, fixed=TRUE), fixed=TRUE), by = .(name, Parent)]

tmp <- strsplit(ICS_data$tmp, "&")
tmp2 <- sapply(tmp, function(x){if(length(x)>1){
  paste(gsub("!", "", gsub("TRUE", "+", gsub("FALSE", "-", paste0(x, !grepl("!", x, fixed=TRUE)), fixed=TRUE), fixed=TRUE), fixed=TRUE), collapse="")
  } else {x}})
ICS_data[, `:=`(tmp = NULL, Population = ifelse(tmp2=="NULL", Population, tmp2))]

# viability
ICS_data[,
         `:=`(CellsAcquired = ParentCount[Population == "SINGLETS"],
              Viability = 100 * Count[Population %in% "LIVE"]/ParentCount[Population %in% "LIVE"]),
         by = name]

# Viability -- Cutpoint specified in protocol_specific.R
ICS_data <- ICS_data[, ViabilityCut := ifelse(Viability < Vpctcut | CellsAcquired < Vrawcut, 'Not Viable', 'Viable')]

# Recompute Marginals from Booleans
ComputedMarginal <- rbindlist(lapply(c("IFNg", "IL2", "IL4", "IL5", "IL13", "IL17A", "IL21"), function(x) {
  ICS_data[Population %like% paste0(x, "\\+") & obstype=="Boolean",
           .(Population = x, Count = sum(Count, na.rm=TRUE), obstype = "Computed Marginal", HierarchyLevel = unique(HierarchyLevel),
             ParentCount = unique(ParentCount), CellsAcquired = unique(CellsAcquired),
             Viability = unique(Viability), ViabilityCut = unique(ViabilityCut)),
           by = c("name", "Parent", kw)]}))

IFNgOrIL2rows <- which(grepl("IFNg+", ICS_data$Population, fixed=TRUE) | grepl("IL2+", ICS_data$Population, fixed=TRUE))
AnyTh2rows <- which(grepl("IL4+", ICS_data$Population, fixed=TRUE) |
                      grepl("IL5+", ICS_data$Population, fixed=TRUE) |
                      grepl("IL13+", ICS_data$Population, fixed=TRUE))

IFNgOrIL2 <- ICS_data[IFNgOrIL2rows,
                      .(Population = "IFNg Or IL2", Count = sum(Count, na.rm=TRUE), obstype = "Computed Joint Marginal",
                        HierarchyLevel = unique(HierarchyLevel), ParentCount = unique(ParentCount),
                        CellsAcquired = unique(CellsAcquired), Viability = unique(Viability), ViabilityCut = unique(ViabilityCut)),
                      by = c("name", "Parent", kw)]

AnyTh2 <- ICS_data[AnyTh2rows,
                   .(Population = "IL4 Or IL5 Or IL13", Count = sum(Count, na.rm=TRUE), obstype = "Computed Joint Marginal",
                     HierarchyLevel = unique(HierarchyLevel), ParentCount = unique(ParentCount),
                     CellsAcquired = unique(CellsAcquired), Viability = unique(Viability), ViabilityCut = unique(ViabilityCut)),
                   by = c("name", "Parent", kw)]

ICS_data <- rbindlist(list(ICS_data, ComputedMarginal, IFNgOrIL2, AnyTh2), fill=TRUE, use.names=TRUE)
print(ICS_data[, .N, obstype])

j <- dcast.data.table(ICS_data[obstype %in% c("Provided Marginal", "Computed Marginal")],
                      name + Population + Parent + `Animal ID` + `Time Point` + Stim ~ obstype,
                      value.var="Count")[!is.na(`Computed Marginal`) & `Computed Marginal` != `Provided Marginal`]

cat("Table of differences between Provided Marginal counts and Computed Marginal Counts\n")
print(j[, .N, .(`Provided Marginal`, `Computed Marginal`)])

cat("Discarding Provided Marginals\n")
ICS_data <- ICS_data[obstype != "Provided Marginal" | Parent == "CD4/CM/CXCR5"]

print(dcast(ICS_data[, .N, .(Stim, Parent, Population, Run)], Parent + Population + Run ~ Stim, value.var="N"), nrow=200)

# background viability
ICS_data[,
         `:=`(CountBG = Count[Stim %like% ICS_control],
              ParentCountBG = ParentCount[Stim %like% ICS_control],
              ViabilityBG = Viability[Stim %like% ICS_control],
              CellsAcquiredBG = CellsAcquired[Stim %like% ICS_control],
              ViabilityCutBG = ViabilityCut[Stim %like% ICS_control]),
         by = .(`Animal ID`, Population, Parent, `Time Point`, obstype, Run)]

# Compute proportions
ICS_data[, `:=`(PercentCell = Count/ParentCount * 100,
                PercentCellNet = (Count/ParentCount - CountBG/ParentCountBG)* 100)]

# fix names
setnames(ICS_data, c("Animal ID", "name", "Time Point"), c("AnimalID", "Name", "Week"))
```

Compute p-values with `fisher.test()` and adjust using `p.adjust()`.

```{r fisher}
# Fisher's test of positive/negative counts for antigen vs. COSTIM
# pops_of_interest is defined in protocol_specific.R
ppop <- ICS_data[!is.na(CountBG) &  paste0(Parent, "/", Population) %in% pops_of_interest &
                   Stim %in% unique(Stim[Stim != ICS_control])  & ViabilityCut == 'Viable' & ViabilityCutBG == 'Viable', which = TRUE]

ICS_data[ppop,
         response_P := fisher.test(matrix(c(Count, CountBG, ParentCount - Count, ParentCountBG - CountBG), nrow = 2),
                                   alternative = 'greater')$p.value,
         by = .(AnimalID, Stim, Population, Parent, Week)]

# Doing Adjustment on Only Viable Values
ICS_data[ppop,
         response_fdr_P := p.adjust(response_P, method = 'fdr'),
         by = .(Stim, Population, Parent, Week)]
```

We now run MIMOSA to estimate responses.

```{r MIMOSA, eval = runMIMOSA}
# MIMOSA probabilities
library(MIMOSA)

wks <- unique(sort(ICS_data$Week))
sts <- setdiff(unique(sort(ICS_data$Stim)), ICS_control)
cyts <- unlist(unique(ICS_data[obstype %in% c("Provided Marginal", "Computed Marginal", "Computed Joint Marginal") &
                                  (Population %like% "IL" | Population %like% "IFNg")][, .(paste0(Parent, "/", Population))]))

for(wk in wks){
  for(st in sts){
    for(cyt in cyts){
      cat(wk, st, cyt, "\n")
      tmp <- ICS_data[Stim %in% c(ICS_control, st) & Week==wk & paste0(Parent, "/", Population)==cyt &
                         ViabilityCut=="Viable" & ViabilityCutBG=="Viable"]
      if(length(unique(tmp$Stim))>1){
        
        tmpOut <- get_MIMOSA_probs_fun(data_in = tmp, ref_antigen_in = ICS_control)
        if(exists("MIMOSA_results")){
          MIMOSA_results <- rbindlist(list(MIMOSA_results,
                                           tmpOut$Results[, `:=`(Week = wk, Stim = st, Seed = tmpOut$seed_used)]))
        } else {
          MIMOSA_results <- tmpOut$Results[, `:=`(Week = wk, Stim = st, Seed = tmpOut$seed_used)]
        }
      }
    }
  }
}

setkey(ICS_data, AnimalID, Stim, Week, Population, Parent)
setkey(MIMOSA_results, AnimalID, Stim, Week, Population, Parent)
ICS_data <- merge(ICS_data, MIMOSA_results, all.x=TRUE)
ICS_data[, response := ifelse(MIMOSA_fdr_P < 0.01, 1, 0)]

# convert factors to character
ICS_data <- setDT(lapply(ICS_data, function(x){if(class(x)=="factor"){x <- as.character(x)} else {x <- x}}))

# sort for output
setkey(ICS_data, AnimalID, Stim, Week, obstype, HierarchyLevel, Parent, Population)

# Output for Package
saveObj("ICS_data", "ICS_data")
```

We now run COMPASS.

```{r COMPASS, eval = runCOMPASS}
# COMPASS CONTAINER

gs <- load_gs(path=file.path(SCRATCH, "gs_booleans"))

# Construct a COMPASS container from single cell gene expression.
# Some markers have inconsistent marker names.
# Method to update markernames
library(COMPASS)

setGeneric("markernames<-", def = function(object, value, ...){standardGeneric("markernames<-")})
setMethod("markernames<-", signature = c("flowFrame", "character"), definition = function (object, value, ...)
{
  .local <- function (object, value,...)
  {
    m <- as.vector(object@parameters$desc)
    ind <- grepl("time|fsc|ssc", object@parameters$name,
                 ignore.case = TRUE)
    m <- m[!ind]
    m[!is.na(m)]
    if(length(m[!is.na(m)])==length(value)){
      object@parameters$desc[!ind][!is.na(m)] <- value
    }else if(length(m)==length(value)){
      object@parameters$desc[!ind] <- value
    }
    object
  }
  .local(object, value, ...)
})

# Need to fix markernames first, as they may not be consistent
newmarkers <- gsub("-", "", str_split_fixed(markernames(flowData(gs)[[1, use.exprs=FALSE]]), " ", 2)[, 1])

fd <- flowData(gs)
for(i in sampleNames(gs)){
  fr <- fd[[i, use.exprs=FALSE]]
  markernames(fr) <- newmarkers
  fd@frames[[i]] <- fr
}
flowData(gs) <- fd

sc <- getSingleCellExpression(gs,
                              nodes = c("cd4/totm/ifn-g",
                                        "cd4/totm/il-2",
                                        "cd4/totm/il-4",
                                        "cd4/totm/il-5",
                                        "cd4/totm/il-13",
                                        "cd4/totm/il-17",
                                        "cd4/totm/il-21"),
                              map = list("cd4/totm/ifn-g" = "IFNg",
                                         "cd4/totm/il-2"  = "IL2",
                                         "cd4/totm/il-4"  = "IL4",
                                         "cd4/totm/il-5"  = "IL5",
                                         "cd4/totm/il-13" = "IL13",
                                         "cd4/totm/il-17" = "IL17A",
                                         "cd4/totm/il-21" = "IL21"))

totals <- getPopStats(gs, subpopulations = "cd4/totm")[, Count]
names(totals) <- getPopStats(gs, subpopulations = "cd4/totm")[, name]
meta <- pData(gs)
meta$name <- rownames(meta)
meta$Stim <- toupper(meta$Stim)

# Construct a COMPASSContainer
ICS_COMPASS <- COMPASSContainer(data = sc, counts = totals, meta = meta, individual_id = "Animal ID", sample_id = "name")

# Combining ID and Time Together for COMPASS Run
ICS_COMPASS$meta$ID_Time <- paste0(ICS_COMPASS$meta$`Animal ID`, '_', ICS_COMPASS$meta$`Time Point`)

# Setting the new individual_id
ICS_COMPASS$individual_id <- 'ID_Time'

# Filter non-viable samples
ics2 <- copy(ICS_data)
nonViable <- ics2[ViabilityCut != "Viable" | ViabilityCutBG != "Viable", which=TRUE]
assign("nonViable", ics2[nonViable, .N, .(ID_time = paste0(AnimalID, "_", Week))]$ID_time, envir=.GlobalEnv)

# Output for Package
saveObj("ICS_COMPASS", "CompassContainer")

# COMPASS run 1: 7 marker run excluding timepoints with problematic IL2
# loop through stim values, one compass run per stim per tissue
scores <- data.table()
fits <- list()

# loop through stim values, one compass run per stim
for(st in Stims){
  assign("thisStimF", st, envir = .GlobalEnv)
  set.seed(93457653)

  # compute fit
  tmp_fit <- COMPASS(data = ICS_COMPASS,
                     treatment = Stim == .GlobalEnv$thisStimF,
                     control = Stim == 'COSTIM',
                     subset = ! ID_Time %in% .GlobalEnv$nonViable,
                     iterations = COMPASS_iterations)

  # functionality and polyfunctionality scores
  FS <- FunctionalityScore(tmp_fit)
  PFS <- PolyfunctionalityScore(tmp_fit)
  tmp_scores <- data.table(ID_Time = names(FS),
                           FS,
                           PFS,
                           Stim = paste(.GlobalEnv$thisStimF, "(CD4+)"))
  
  tmp_scores <- merge(setDT(tmp_fit$data$meta)[, .(ID_Time, `Animal ID`, `Time Point`, Group)],
                      tmp_scores, by = "ID_Time")
  
  # append this scores dataset to the master scores dataset
  scores <- rbindlist(list(scores, tmp_scores), use.names = TRUE, fill = TRUE)

  # save the stim- and tissue-specific fit stats to permanent dataset
  fits[[st]] <- tmp_fit
}

# Output for Package
saveObj("scores", "ICS_scores")
saveObj("fits", "ICS_fits")
```

```{r PLOTS, eval=FALSE}
# Create plots for samples with high cell counts

# get samples having abnormally high counts
samples <- unique(ICS_data[Count>500 & Parent == "CD4/TOTM" & obstype %like% "Marginal"]$Name)

gs <- load_gs(path=file.path(SCRATCH, "gs_booleans"))

# standardize case of Stim in gating set hierarchy
pd <- pData(gs)
pd$Stim <- ifelse(pd$Stim==ICS_control, pd$Stim, toupper(pd$Stim))
pData(gs) <- pd

pd$name <- rownames(pd)
pd <- setnames(setDT(pd)[, `:=`(Stim = ifelse(Stim==ICS_control, Stim, toupper(Stim)))], "Animal ID", "AnimalID")

highCounts <- pd[name %in% samples]
highCountsBG <- merge(copy(highCounts)[, c("Stim", "name") := NULL], pd[Stim==ICS_control, .(AnimalID, `Time Point`, Run, Stim, name)], by=c("AnimalID", "Time Point", "Run"))
highCountsBG <- highCountsBG[!duplicated(highCountsBG)]
sample_list <- rbindlist(list(highCounts, highCountsBG), use.names=TRUE, fill=TRUE)

sample_map <- sample_list[, .(AnimalID, `Time Point`, Stim, id=interaction(AnimalID, `Time Point`), name)]
sample_map <- setkey(sample_map[!duplicated(sample_map)], id)
sample_map2 <- lapply(unique(sample_map$id), function(x){c(sample_map[id==x]$name)})

nodesToPlot <- c("/singlets/live/cd3/cd4/totm/ifn-g", "/singlets/live/cd3/cd4/totm/il-2")

quartz()

for(i in sample_map2){
  for(node in nodesToPlot){
    sp <- strsplit(node, "/")[[1]]
    fn <- gsub("\\.fcs_[0-9]*", "", paste0(gsub("Specimen_", "", i), "-", paste(sp[5:length(sp)], collapse="."), ".png"))
    png(file.path("..", "vignettes", "plots", fn))
    print(plotGate(gs[i], node, cond="Stim + `Time Point` + `Animal ID`", path = "auto",
                   marker.only=TRUE, overlay=c("cd4/totm/il-2","cd4/totm/ifn-g"),
                   par.settings = list(overlay.symbol = list(cex = 0.4)), xbin=128, margin=FALSE, main = node))
    dev.off()
  }
}
```

## Reproducibility Tables

```{r echo=FALSE, eval=TRUE}
path <- "reproducibility.R")
read_chunk(path)
```

```{r repro, echo=FALSE, eval=TRUE}
<<reprotab>>
```

```{r copySourcedFiles, eval = FALSE, echo = FALSE}
# this copies any files that are sourced in this Rmd to inst/doc/ and vignettes/ so that the package vignettes build properly at installation
# this is new code, and may still be a little buggy or not function as expectged, particularly on Windows
thisfile <- knitr::current_input()
con <- file(thisfile, open = "r")
file <- readLines(con)
m1 <- unname(sapply(file, function(x){grepl('sys.source(.*)', x) & !grepl('unname', x)}))
m2 <- unname(sapply(file, function(x){grepl('read_chunk(.*)', x) & !grepl('unname', x)}))
filesToCopy1 <- sapply(strsplit(file[m1], split = '"'), function(x) x[2])
filesToCopy2 <- sapply(strsplit(file[m2], split = '"'), function(x) x[2])
filesToCopy <- c(filesToCopy1, filesToCopy2)
for(f in filesToCopy){
  file_copy(f, paste0(c("../inst/doc/", "../vignettes/"), f), overwrite = TRUE)
}
```

